### PR TITLE
Initialize service isolate from kernel file, rather than from embedded sources.

### PR DESCRIPTION
**Pull request by aam**
Currently service isolate is initialized from the source code parsed by VM.
This CL changes it so service isolate created during Dart initialization
is created from the kernel platform.dill file if it is present in the application
bundle. Then this platform kernel file is kept in dart_init module and reused
for application sciprt isolates.

### CODE REVIEW

File: runtime/dart_init.cc [hunk 1]
```diff
@@ -194,10 +199,19 @@ Dart_Isolate ServiceIsolateCreateCallback(const char* script_uri,
   return nullptr;
 #else   // FLUTTER_RUNTIME_MODE
   UIDartState* dart_state = new UIDartState(nullptr, nullptr);
+
+  bool is_running_from_kernel = kernel_platform != nullptr;
```
Comment (a-siva) [inline runtime/dart_init.cc]: Should this be GetKernelPlatformBinary() != nullptr to preserve the abstraction?

File: runtime/dart_init.cc [hunk 2]
```diff
@@ -216,12 +230,14 @@ Dart_Isolate ServiceIsolateCreateCallback(const char* script_uri,
       const bool disable_websocket_origin_check = false;
       const bool service_isolate_booted = DartServiceIsolate::Startup(
           ip, port, tonic::DartState::HandleLibraryTag,
-          IsRunningPrecompiledCode(), disable_websocket_origin_check, error);
+          IsRunningPrecompiledCode() || is_running_from_kernel,
```
Comment (a-siva) [inline runtime/dart_init.cc]: This is going to be confusing inside DartServiceIsolate::Startup method as the parameter there states 'running_precompiled' which won' be true in this case.

Maybe the parameter 'running_precompiled' should be changed to 'running_fromsource' ?

File: runtime/dart_init.cc (hunk 1 above)
Comment (aam) [reply runtime/dart_init.cc]: Done.

File: shell/common/diagnostic/diagnostic_server.cc [hunk 3]
```diff
@@ -72,16 +74,22 @@ void DiagnosticServer::Start(uint32_t port, bool ipv6) {
...
-      resources.ResourceLookup(kDiagnosticServerScript, &source);
-  FXL_DCHECK(source_length != EmbedderResources::kNoSuchInstance);
-
-  Dart_Handle diagnostic_library = Dart_LoadLibrary(
-      Dart_NewStringFromCString("dart:diagnostic_server"), Dart_Null(),
-      Dart_NewStringFromUTF8(reinterpret_cast<const uint8_t*>(source),
-                             source_length),
-      0, 0);
+  Dart_Handle diagnostic_library;
+  if (running_from_kernel) {
```
Comment (a-siva) [inline shell/common/diagnostic/diagnostic_server.cc]: Why is it necessary to wire it this way using a parameter, is it not sufficient for this check to be
if (blink::GetKernelPlatformBinary() != nullptr) {
  ....

File: runtime/dart_init.cc [hunk 4]
```diff
@@ -124,6 +124,11 @@ static ServiceIsolateHook g_service_isolate_hook = nullptr;
 static RegisterNativeServiceProtocolExtensionHook
     g_register_native_service_protocol_extensions_hook = nullptr;
 
+// Kernel representation of core dart libraries(loaded from platform.dill).
+static void* kernel_platform = nullptr;
```
Comment (a-siva) [inline runtime/dart_init.cc]: This value once set will not be deleted anymore? I don't see an engine teardown path for us to delete this, maybe something to think about for another CL (We could also call Dart_Cleanup there which I noticed was not being called anywhere).

File: shell/common/diagnostic/diagnostic_server.cc (hunk 3 above)
Comment (aam) [reply shell/common/diagnostic/diagnostic_server.cc]: Done

File: runtime/dart_init.cc (hunk 2 above)
Comment (aam) [reply runtime/dart_init.cc]: Done

File: runtime/dart_init.cc (hunk 4 above)
Comment (aam) [reply runtime/dart_init.cc]: Added TODO.

**Comment by pylaligand** [comment]
This broke the Fuchsia build:
```
../../third_party/flutter/content_handler/runtime_holder.cc:166:54: error: too few arguments to function call, expected 5, have 4
                       default_isolate_snapshot_instr);
                                                     ^
../../third_party/flutter/runtime/runtime_init.h:13:1: note: 'InitRuntime' declared here
void InitRuntime(const uint8_t* vm_snapshot_data,
^
1 error generated.
```

**Comment by aam** [comment]
Sorry about that. https://github.com/flutter/engine/pull/4274 adds missing parameter.

File: shell/common/engine.cc [hunk 5]
```diff
@@ -248,15 +244,9 @@ void Engine::RunBundleAndSource(const std::string& bundle_path,
   if (packages_path.empty())
     packages_path = FindPackagesPath(main);
 
-  std::vector<uint8_t> platform_kernel;
-  if (!bundle_path.empty()) {
-    ConfigureAssetBundle(bundle_path);
```
Comment (xster) [inline shell/common/engine.cc]: Wonder if this causes the same issue as https://github.com/flutter/engine/pull/4219. Seems to cause a regression https://github.com/flutter/flutter/issues/12739 that we've seen before.

Comment (aam) [reply shell/common/engine.cc]: I'm sure it does! Sorry about that.

Comment (aam) [reply shell/common/engine.cc]: https://github.com/flutter/engine/pull/4285 with the fix.

**Pull request merged**